tenpy/networks/site.py

```diff
@@ -1906,3 +1907,85 @@ def spin_half_species(SpeciesSite, cons_N, cons_Sz, **kwargs):
         new_mod.append(4)
     set_common_charges([up_site, down_site], new_charges, new_names, new_mod)
     return [up_site, down_site], ['up', 'down']
+
+
+class ClockSite(Site):
+    r"""Quantum clock site.
+
+    There are ``q`` local states, with labels ``['0', '1', ..., str(q-1)]``.
+    Special aliases are ``up`` (0), and if q is even ``down`` (q / 2).
+    Local operators are the clock operators ``Z = diag([w ** 0, w ** 1, ..., w ** (q - 1)])``
+    with ``w = exp(2.j * pi / q)`` and ``X = eye(q, k=1) + eye(q, k=1-q)``, which are not hermitian (!)
+
+    =========================== ================================================
+    operator                    description
+    =========================== ================================================
+    ``Id, JW``                  Identity :math:`\mathbb{1}`
+    ``X, Z``                    Clock operators
+    ``Xhc, Zhc``                Hermitian conjugates of clock operators
+    ``Xphc, Zphc``              Clock operator plus its hermitian conjugate
+    =========================== ================================================
+
+    ============== ====  ============================
+    `conserve`     qmod  *excluded* onsite operators
+    ============== ====  ============================
+    ``'Z'``        [q]   ``Xphc, Zphc``
+    ``'None'``     []    --
+    ============== ====  ============================
+
+    Parameters
+    ----------
+    q : int
+        Number of states per site
+    conserve : str | None
+        Defines what is conserved, see table above.
+    sort_charge : bool
+        Whether :meth:`sort_charge` should be called at the end of initialization.
+        This is usually a good idea to reduce potential overhead when using charge conservation.
+        Note that this permutes the order of the local basis states!
+        For backwards compatibility with existing data, it is not (yet) enabled by default.
+
+    Attributes
+    ----------
+    q : int
+        Number of states per site
+    conserve : str
+        Defines what is conserved, see table above.
+    """
+    def __init__(self, q, conserve='Z', sort_charge=None):
+        if not (isinstance(q, int) and q > 1):
+            raise ValueError(f'invalid q: {q}')
+        self.q = q
+        if not conserve:
+            conserve = 'None'
+        if conserve not in ['Z', 'None']:
+            raise ValueError("invalid `conserve`: " + repr(conserve))
+        X = np.eye(q, k=1) + np.eye(q, k=1-q)
+        Z = np.diag(np.exp(2.j * np.pi * np.arange(q, dtype=np.complex128) / q))
+        Xhc = X.conj().transpose()
+        Zhc = Z.conj().transpose()
+        Xphc = X + Xhc
+        Zphc = np.diag(2. * np.cos(2. * np.pi * np.arange(q, dtype=np.complex128) / q))
+        if conserve == 'Z':
+            # we store n as the charge where <Z> = exp(2.j * pi * n / q)
+            # TODO (JU): document this somewhere?
```

Comment:
If in doubt, one needs to check the code here. It's pretty straight-forward to find in the code, so I think there's no need to document it in the doc string of the site, similar as for the other sites.

Reply:
ok